src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs

```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+            if (documentNames == null)
+            {
+                return false;
+            }
+
+            // Write out the documents.
+            Directory.CreateDirectory(context.OutputDirectory);
+            var filePathList = new List<string>();
+            foreach (var documentName in documentNames)
+            {
+                var filePath = GetDocument(
+                    documentName,
+                    context.ProjectName,
+                    context.OutputDirectory,
+                    generateMethod,
+                    service);
+                if (filePath == null)
                 {
-                    Reporter.WriteWarning(Resources.FormatMethodReturnTypeUnsupported(
-                        methodName,
-                        serviceName,
-                        method.ReturnType,
-                        typeof(Task)));
                     return false;
                 }
 
-                var service = services.GetService(serviceType);
-                if (service == null)
+                filePathList.Add(filePath);
+            }
+
+            // Write out the cache file.
+            var stream = File.Create(context.FileListPath);
+            using var writer = new StreamWriter(stream) { AutoFlush = true };
+            writer.WriteLine(string.Join(Environment.NewLine, filePathList));
+            stream.Flush();
+
+            return true;
+        }
+
+        private static string GetDocument(
+            string documentName,
+            string projectName,
+            string outputDirectory,
+            MethodInfo generateMethod,
+            object service)
+        {
+            Reporter.WriteInformation(Resources.FormatRetrievingDocument(documentName));
+
+            var stream = new MemoryStream();
+            using var writer = new StreamWriter(stream) { AutoFlush = true };
+            var resultTask = (Task)InvokeMethod(generateMethod, service, new object[] { documentName, writer });
+            if (resultTask == null)
+            {
+                return null;
+            }
+
+            var finishedIndex = Task.WaitAny(resultTask, Task.Delay(TimeSpan.FromMinutes(1)));
+            if (finishedIndex != 0)
+            {
+                Reporter.WriteError(Resources.FormatMethodTimedOut(GenerateMethodName, DocumentService, 1));
+                return null;
+            }
+
+            var filePath = GetDocumentPath(documentName, projectName, outputDirectory);
+            try
+            {
+                if (stream.Length == 0L)
                 {
-                    Reporter.WriteWarning(Resources.FormatServiceNotFound(serviceName));
-                    return false;
+                    Reporter.WriteError(Resources.FormatMethodWroteNoContent(
+                        GenerateMethodName,
+                        DocumentService,
+                        documentName));
+
+                    return null;
```

Comment:
Is there a reason for this to be inside the `try/catch`? This seems like it belongs higher up.